iree_special_models/sd3/mmdit/test_mmdit.py

```diff
@@ -0,0 +1,120 @@
+# Copyright 2024 Advanced Micro Devices, Inc.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import pytest
+from ireers import *
+import os
+
+# iree_special_models/sd3/mmdit
+current_dir = os.path.dirname(os.path.realpath(__file__))
+iree_test_path_extension = os.getenv("IREE_TEST_PATH_EXTENSION", default=current_dir)
+rocm_chip = os.getenv("ROCM_CHIP", default="gfx90a")
+
+###############################################################################
+# Fixtures
+###############################################################################
+
+CPU_COMPILE_FLAGS = [
+    "--iree-hal-target-backends=llvm-cpu",
+    "--iree-llvmcpu-target-cpu-features=host",
+    "--iree-llvmcpu-fail-on-out-of-bounds-stack-allocation=false",
+    "--iree-llvmcpu-distribution-size=32",
+    "--iree-opt-const-eval=false",
+    "--iree-llvmcpu-enable-ukernels=all",
+    "--iree-global-opt-enable-quantized-matmul-reassociation"
+]
```

Comment:
(Sorry, thoughts bouncing around here :P)

I guess for testing in IREE we'd consider forking this entire file so these options could be overridden and new configurations could be added (e.g. Vulkan, CUDA), but at that point we'd be just sharing
* the 90 lines from ireers/fixtures.py
* The method for downloading files

So I'm wondering if it would make more sense to fold changes into https://github.com/iree-org/iree/tree/main/experimental/regression_suite and get that out of `experimental/`. The file downloading _is_ unique to nod-ai/SHARK-TestSuite though... IREE upstream would prefer to directly use some neutral hosted location instead of the shark Azure buckets.

... (insert thinking noises here) ...

How about this: we keep this PR mostly as-is and treat it as a "community CI build" in IREE. AMD is supplying the runners, data, and test files, and all of that configuration exists here in this repo. As the move into the Linux Foundation progresses and this test suite gets some mileage we can consider moving it to iree-org/, using some neutral cloud hosting, and including models (and compile flags, etc.) that other contributors care about.

It _would_ still be useful to be able to change e.g. `ROCM_COMPILE_FLAGS` from IREE though - especially if a flag is renamed there... we wouldn't want the test to fail with no way to fix it that doesn't cross repos.

Thoughts?

Reply:
Yeah your thoughts are basically what I was thinking too. It would be good to have the flags changeable in IREE. My main thing was I wanted to eliminate all those json files with this implementation, so I just went with this. The one way I thought was maybe having a special_models common json file. We would add all the compile flags for all the models. But, then I thought this would be too messy as we add models (can still do that, just not sure how sustainable/neccessary it is). If we split it up based on model, it would just end up like the original situation with a bunch of json files.  I think the main thing is I couldn't think of a sustainable way where we would not have to add a bunch of files. Open for any option that doesn't require a bunch of files though.

Reply:
Yeah this is definitely an improvement. Forking the .py files (to files like https://github.com/iree-org/iree/blob/main/experimental/regression_suite/tests/pregenerated/test_llama2.py) may be reasonable. What I'm watching out for is:

* IREE contributors without access to these Azure storage accounts not getting stuck if the files need to be regenerated (so need a way to either contribute updates or at least disable tests temporarily)
* IREE contributors need a way to change flags and configurations (add new ones for testing, remove flags after changing them, etc.)
* Complexity for:
  * Adding new tests
  * Seeing which tests are defined
  * Seeing which flags are used
  * Checking which tests are passing/failing
  * Adding benchmarks

We're talking about ~300 lines of code per test model (across multiple configurations), so ~5000 lines of code total across ~15 test models. That's not _too_ bad for maintenance.

> The one way I thought was maybe having a special_models common json file. We would add all the compile flags for all the models on cpu and rocm.

Actually for this, we can have common python files, so that might not be so bad.

```python
DEFAULT_ROCM_FLAGS = [ "foo", "bar" ]
WINOGRAD_ROCM_FLAGS = ...
SDXL_ROCM_FLAGS = DEFAULT_ROCM_FLAGS + ...

ALL_BACKENDS = ...

def register_tests_for_model(model=..., excluded_backends=[...])
```

if we add that sort of configuration then it should really live in the IREE repo and not here. What we _could_ store here would be model importing tools, sanity check tests, git LFS files, or URLs to huggingface / Azure / GCS / etc.

Reply:
Cool, this sounds good to me. Can we have the test files live here and in iree? I think there is a value to be able to run all tests out of here (e2e, iree normal models, iree special models). IREE can use the TestSuite and have its own test files so easy to make changes there, but I would like the TestSuite to also be something that you can clone and be able to run tests on everything SHARK related.

Reply:
Yeah, I guess let's be pragmatic. We can keep a copy here and in IREE and just sync them manually as needed.

If we have only one place exhaustively test configurations, I'd prefer for that to be IREE (to catch regressions and measure performance using code on PRs and landed commits).

Reply:
Ok yeah makes sense